### ISSUE TITLE
Create 2 jobs to fetch Primero case data and map and load to Child Shield 

### ISSUE THREAD

**Issue by daissatou2** [OpenFn/unicef-thailand#30]
## Background, context, and business value

Child Shield is an AI software developed by the MOPH that ingests child health and protection data to determine what the child is at risk of. In Flow 1, we will be syncing Primero case data to the Child Shield system.

### **JOB 1: Get Primero Cases**
1. Fetch cases from Primero where `last_updated_at` is >= cursor date. If no cases are found, the job should log the message "No cases updated."
2. If cases are found, follow the steps below to filter out cases that do not meet the valid criteria. The output will be a list of "valid cases" to be sent to Child Shield.

3. **Syncing closed cases**
For each case, where `status` = "closed"...
-- Is `assessment_requested_on` == null? If yes, remove the case from the list of valid cases.
-- If no, is `date_closure`  >= cursor date? If no, remove the case from the list of valid cases.

4. **Syncing open cases**
For cases where `status` != "closed"...
-- Is `assessment_requested_on`  >= cursor date? If no, remove the case from the list of valid cases. 

Log the final array of valid cases---they should be passed on as the initial state for Job 2. 

### **JOB 2: Get translations & output mapping JSON object** 
This job was delivered for issue #31. It will need to be changed to a "Flow" trigger job... that will run after Job 1 successfully completed. 

Its final state should include both (1) the translations JSON mapping object (what was delivered in #31) _AND_ (2) the cases extract in the prior Job 1. 
See this diagram for how the final state should be passed between these jobs: https://lucid.app/lucidchart/b8ee8b9d-2f0f-4d19-b12e-908d4138ccb4/edit?viewport_loc=-270%2C-26%2C3117%2C1500%2CnEyQdSFm_Akd&invitationId=inv_093b1b38-48bc-421c-a25e-0fbb5a1b7a8d#

### **JOB 3: Map & Load Interventions to Child Shield** 
 See issue #47 for these details. 

**Note**: We do not have the full JSON path for each field being mapped to Child Shield; we only have the field name or "matching attribute" (column c in mapping spec). For each field, this job should search through the inbound payload from job 1 and grab the value for the matching attribute. See slack [thread](https://openfn.slack.com/archives/C017ELVRSM8/p1660596219397849) and  TD's suggested code below. 

``` js
function lookup(obj, k) {
  for (var key in obj) {
    var value = obj[key];

    if (k == key) {
      return [k, value];
    }

    if (typeof(value) === "object" && !Array.isArray(value)) {
      var y = lookup(value, k);
      if (y && y[0] == k) return y;
    }
    if (Array.isArray(value)) {
      // for..in doesn't work the way you want on arrays in some browsers
      //
      for (var i = 0; i < value.length; ++i) {
        var x = lookup(value[i], k);
        if (x && x[0] == k) return x;
      }
    }
  }

  return null;
}
```

```js
function type(object) {
  var stringConstructor = "test".constructor;
  var arrayConstructor = [].constructor;
  var objectConstructor = {}.constructor;

  if (object === null) {
    return "null";
  } else if (object === undefined) {
    return "undefined";
  } else if (object.constructor === stringConstructor) {
    return "String";
  } else if (object.constructor === arrayConstructor) {
    return "Array";
  } else if (object.constructor === objectConstructor) {
    return "Object";
  } else {
    return "null";
  }
}
```

[**DATA FLOW DIAGRAM**](https://lucid.app/lucidchart/b8ee8b9d-2f0f-4d19-b12e-908d4138ccb4/edit?beaconFlowId=64F72BCB10CA89B0&invitationId=inv_093b1b38-48bc-421c-a25e-0fbb5a1b7a8d&page=gz2GymXUIyeW#)

<img width="1161" alt="image" src="https://user-images.githubusercontent.com/80456839/191881642-22c9a66c-b65e-41aa-a270-bdcf676b4918.png">

## state.json

JOB 1: 
```json
{
  "configuration": { "SEE LP Primero Thailand Demo API User (Open Function)" },
  "cursor": "Fetch cases where "last_updated_at" >= date of last sync"
}
```

JOB 2:
```json
{
  "configuration": ["See LP MOPH Thailand HIS API - DEMO Alpha (Unicef)'"],
  "data": { "data generated by GET job"},
}
```

## adaptor
Job 1: language-primero 
Job 2: language-http

[Primero API docs](https://github.com/primeroIMS/primero/tree/development_v2/app/controllers/api)
[Child Shield API docs](https://cloud1-r8way.moph.go.th:3011/explorer/#!/intervention/intervention_create)
[Child Shield Sample Postman Requests](https://docs.google.com/document/d/1GIRxVXSZLgmKwTf-dP0kQjto5-VdZl5GjCj-pAjKfaA/edit#)

## trigger
Job 1: Cron
Job 2: Flow

## expression.js
Add both jobs to a new child-shield directory in the repo. 

Job 1: Fetch Primero cases according to the criteria above. Filter out irrelevant cases. Output remaining cases.
Job 2: Map and load remaining cases to Child Shield.

## Toggl 
`UNICEF Thailand`

**Comment by aleksa-krolls** [comment]
@mtuchi Jobs 1 and 2 are ready to start on. *Please hold on job 3 until we get confirmation from the client.* To discuss shortly in our call. Pls let @daissatou2 know if you have any questions. 

For Job 1 - [here is an example Job](https://github.com/OpenFn/primero-progres/blob/master/jobs/PROD/f2-1.getCasesFromPrimero.js#L11-L16) from another project where we "getCases" from Primero, and search only for cases that were `last_updated_at` since the cursor date ([see L5 here](https://github.com/OpenFn/primero-progres/blob/master/jobs/PROD/f2-1.getCasesFromPrimero.js#L5)), which is _either_ the job's last runtime OR a default date that we choose. 

Your code will probably look something like...
```
fn(state => {
  console.log('Last job sync runtime:', state.lastRunDateTime);
  const manualCursor = '2022-05-01T00:00:00.000Z';
  const cursor = state.lastRunDateTime || manualCursor;
  return { ...state, cursor };
});

getCases(
    {
      remote: true,
      last_updated_at: `${state.cursor}..`, //notice the '..' after the cursor; in primero this will filter for data updated SINCE then
    }, state => { 
    console.log('Data extracted from Primero ::`, state.data); 
    return state;
    }); 
```
And then you'll have to follow Aicha's data flow diagram to then process this case data and filter the list of cases returned even further.

**Comment by daissatou2** [comment]
@mtuchi please take a look at my feedback below for JOB 1. Let's discuss any further questions during our sync tomorrow. 

1. It doesn’t look like the job is currently updating the cursor **_before_** the end of each run. So the next time it runs, it doesn’t only fetch cases updated since the last sync, but always fetches cases updated since the “manual cursor”.  See this [code](https://github.com/OpenFn/primero-progres/blob/85130939774d2eb5e8a18d4efc71bc813be5bfcf/jobs/PROD/f2-1.getCasesFromPrimero.js#L55) **for an example of how to update the cursor each time the job runs successfully.** For example, if I run the job at 12:20 pm today and get 5 cases returned from Primero, then I run it again at 12:22 pm (and no cases have been updated in Primero within those 2 minutes) I should see 0 cases returned from Primero, not 5 again. 
2. Open cases: In this [run log](https://www.openfn.org/projects/primero-thailand-io-unicef---staging/runs/063234ac-3d92-7f10-8700-d8003184c0af) a case with `assessment_requested_on: '2022-09-15'` was filtered out of valid cases. Why? It should be considered a valid case because the cursor date is set to `2022-09-15T09:43:07.000Z` in the job. 
3. Closed cases: The case with `case_id: '5f90b60d-a8e9-4256-b207-84b5eeaf189d'` in this [run log](https://www.openfn.org/projects/primero-thailand-io-unicef---staging/runs/063234e9-963e-7ced-9102-e698683f7d18
) has a value for `“assessment_requested_on”` AND has a `date_closure: '2022-09-15'` which is >= the cursor date `(2022-09-15T09:43:07.000Z)`. It should not have been filtered out of the valid cases. 

@aleksa-krolls thoughts on numbers 2 & 3 above? Does that sound right?

**Comment by aleksa-krolls** [comment]
@daissatou2 I agree that these cases shouldn't been filtered out. 
@mtuchi for 2 and 3, when we compare dates, can we compare only the *day* if time is not available...

e.g., if `assessment_requested_on: '2022-09-15'` and cursor date is set to `2022-09-15T09:43:07.000Z`, then
compare is `2022-09-15` >= ``2022-09-15` (without the timestamp)

**Comment by mtuchi** [comment]
> @daissatou2 I agree that these cases shouldn't been filtered out. @mtuchi for 2 and 3, when we compare dates, can we compare only the _day_ if time is not available...
> 
> e.g., if `assessment_requested_on: '2022-09-15'` and cursor date is set to `2022-09-15T09:43:07.000Z`, then compare is `2022-09-15` >= ``2022-09-15` (without the timestamp)

This has been resolved, on this PR https://github.com/OpenFn/primero-thailand/pull/39

To properly test on platform, you will need to use set  `cursor = manualCursor`
On line 6 -> https://github.com/OpenFn/primero-thailand/blob/master/jobs/child-shield/getCasesFromPrimero.js
Change 👇🏽  
```
 const cursor =
    state.lastRunDateTime != null && state.lastRunDateTime != ''
      ? state.lastRunDateTime
      : manualCursor;
```
To 
```
 const cursor = manualCursor;
```

The current value of  `manualCursor  = '2022-09-15T06:43:07.000Z';` [on line 4](https://github.com/OpenFn/primero-thailand/blob/master/jobs/child-shield/getCasesFromPrimero.js)

⚠️  **Don't forget to revert back the changes once your done with the test** 

cc @daissatou2 @aleksa-krolls

**Comment by daissatou2** [comment]
@mtuchi thanks! I tested and Job 1 is looking great. FYI instead of updating the job code, I cleared the job state by clicking the little trash can icon. After clearing the state the job will use the manual cursor on the next run. 
<img width="1008" alt="image" src="https://user-images.githubusercontent.com/80456839/191142400-5c4b10af-503d-4953-b54f-fadaf55ce2b9.png">


I will leave this issue in review as we wait for API updates from the UNICEF team for job 3 (hopefully by tomorrow). Once it is ready for you, I will leave a comment and move to backlog.

**Comment by daissatou2** [comment]
@aleksa-krolls please review what we have for job 3 before we hand off to Mtuchi. I have left you comments in the mapping specifications document. Note: POST intervention has not been tested yet and on #5 (job 3), we’ll need to clarify what it looks like to not have an intervention. Will we get a 404 response, for example? This is pending Tip’s response.

**Comment by aleksa-krolls** [comment]
@daissatou2 as discussed: 
1. I have updated the data flow diagram to show how we filter for 'oscc' interventions
2. Please update the spec & our new API Docs to explain how the request body will differ for `POST` vs. `PATCH` intervention scenarios. As discussed, the form mappings for individual questions look OK... but up until the forms content, we want to describe how the payload should be built. 

Lmk when done, and I'll review this issue again before submitting to Mtuchi.

**Comment by daissatou2** [comment]
@aleksa-krolls this is ready for review.

**Comment by aleksa-krolls** [comment]
@daissatou2 I moved the job#3 specs to separate issue #47. Therefore, can we close this one out?
